File: internal/mcpserver/forwards.go
package mcpserver

import (
	"context"

	"github.com/martinzachariassen/domeneshop-mcp/internal/domeneshop"
	"github.com/modelcontextprotocol/go-sdk/mcp"
)

func registerForwardTools(s *mcp.Server, client *domeneshop.Client) {
	type listForwardsArgs struct {
		DomainID int `json:"domain_id" jsonschema:"the Domeneshop domain ID, from list_domains"`
	}
	mcp.AddTool(s, &mcp.Tool{
		Name:        "list_forwards",
		Description: "List all HTTP forwards configured for a domain.",
		Annotations: readOnly(),
	}, func(ctx context.Context, req *mcp.CallToolRequest, args listForwardsArgs) (*mcp.CallToolResult, any, error) {
		forwards, err := client.ListForwards(ctx, args.DomainID)
		if err != nil {
			return nil, nil, err
		}
		return jsonResult(forwards)
	})

	type getForwardArgs struct {
		DomainID int    `json:"domain_id" jsonschema:"the Domeneshop domain ID, from list_domains"`
		Host     string `json:"host" jsonschema:"the host/subdomain the forward applies to, e.g. 'www' or '@' for the root"`
	}
	mcp.AddTool(s, &mcp.Tool{
		Name:        "get_forward",
		Description: "Get the HTTP forward configured for a specific host.",
		Annotations: readOnly(),
	}, func(ctx context.Context, req *mcp.CallToolRequest, args getForwardArgs) (*mcp.CallToolResult, any, error) {
		forward, err := client.GetForward(ctx, args.DomainID, args.Host)
		if err != nil {
			return nil, nil, err
		}
		return jsonResult(forward)
	})

	type forwardArgs struct {
		DomainID int    `json:"domain_id" jsonschema:"the Domeneshop domain ID, from list_domains"`
		Host     string `json:"host" jsonschema:"the host/subdomain to forward, e.g. 'www' or '@' for the root"`
		URL      string `json:"url" jsonschema:"the destination URL to forward requests to"`
		Frame    bool   `json:"frame,omitempty" jsonschema:"if true, show the destination in a frame instead of redirecting"`
	}
	mcp.AddTool(s, &mcp.Tool{
		Name:        "create_forward",
		Description: "Create a new HTTP forward for a domain. Fails if the host already has a forward or a conflicting A, AAAA, ANAME or CNAME record.",
		Annotations: additive(),
	}, func(ctx context.Context, req *mcp.CallToolRequest, args forwardArgs) (*mcp.CallToolResult, any, error) {
		created, err := client.CreateForward(ctx, args.DomainID, domeneshop.Forward{Host: args.Host, URL: args.URL, Frame: args.Frame})
		if err != nil {
			return nil, nil, err
		}
		return jsonResult(created)
	})

	mcp.AddTool(s, &mcp.Tool{
		Name:        "update_forward",
		Description: "Update an existing HTTP forward for a domain.",
		Annotations: destructive(),
	}, func(ctx context.Context, req *mcp.CallToolRequest, args forwardArgs) (*mcp.CallToolResult, any, error) {
		updated, err := client.UpdateForward(ctx, args.DomainID, args.Host, domeneshop.Forward{Host: args.Host, URL: args.URL, Frame: args.Frame})
		if err != nil {
			return nil, nil, err
		}
		return jsonResult(updated)
	})

	type deleteForwardArgs struct {
		DomainID int    `json:"domain_id" jsonschema:"the Domeneshop domain ID, from list_domains"`
		Host     string `json:"host" jsonschema:"the host/subdomain whose forward should be deleted"`
	}
	mcp.AddTool(s, &mcp.Tool{
		Name:        "delete_forward",
		Description: "Delete an HTTP forward.",
		Annotations: destructive(),
	}, func(ctx context.Context, req *mcp.CallToolRequest, args deleteForwardArgs) (*mcp.CallToolResult, any, error) {
		if err := client.DeleteForward(ctx, args.DomainID, args.Host); err != nil {
			return nil, nil, err
		}
		return textResult("Forward deleted.")
	})
}

File: src/mcp/tools/forwards.ts
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import type { DomeneshopClient } from "../../domeneshop/index.js";
import { additive, destructive, readOnly } from "../annotations.js";
import { jsonResult, textResult } from "../result.js";
import type { InferShape } from "../schema.js";
import { toolHandler } from "../tool.js";

const listForwardsShape = {
  domain_id: z.number().int().describe("the Domeneshop domain ID, from list_domains"),
};
type ListForwardsArgs = InferShape<typeof listForwardsShape>;

const getForwardShape = {
  domain_id: z.number().int().describe("the Domeneshop domain ID, from list_domains"),
  host: z
    .string()
    .describe("the host/subdomain the forward applies to, e.g. 'www' or '@' for the root"),
};
type GetForwardArgs = InferShape<typeof getForwardShape>;

const forwardShape = {
  domain_id: z.number().int().describe("the Domeneshop domain ID, from list_domains"),
  host: z.string().describe("the host/subdomain to forward, e.g. 'www' or '@' for the root"),
  url: z.string().describe("the destination URL to forward requests to"),
  frame: z
    .boolean()
    .optional()
    .describe("if true, show the destination in a frame instead of redirecting"),
};
type ForwardArgs = InferShape<typeof forwardShape>;

const deleteForwardShape = {
  domain_id: z.number().int().describe("the Domeneshop domain ID, from list_domains"),
  host: z.string().describe("the host/subdomain whose forward should be deleted"),
};
type DeleteForwardArgs = InferShape<typeof deleteForwardShape>;

export function registerForwardTools(server: McpServer, client: DomeneshopClient): void {
  server.registerTool(
    "list_forwards",
    {
      description: "List all HTTP forwards configured for a domain.",
      inputSchema: listForwardsShape,
      annotations: readOnly(),
    },
    toolHandler<ListForwardsArgs>(async ({ domain_id }) => {
      const forwards = await client.forwards.list(domain_id);
      return jsonResult(forwards);
    }),
  );

  server.registerTool(
    "get_forward",
    {
      description: "Get the HTTP forward configured for a specific host.",
      inputSchema: getForwardShape,
      annotations: readOnly(),
    },
    toolHandler<GetForwardArgs>(async ({ domain_id, host }) => {
      const forward = await client.forwards.get(domain_id, host);
      return jsonResult(forward);
    }),
  );

  server.registerTool(
    "create_forward",
    {
      description:
        "Create a new HTTP forward for a domain. Fails if the host already has a forward or a conflicting A, AAAA, ANAME or CNAME record.",
      inputSchema: forwardShape,
      annotations: additive(),
    },
    toolHandler<ForwardArgs>(async ({ domain_id, host, url, frame }) => {
      const created = await client.forwards.create(domain_id, { host, url, frame: frame ?? false });
      return jsonResult(created);
    }),
  );

  server.registerTool(
    "update_forward",
    {
      description: "Update an existing HTTP forward for a domain.",
      inputSchema: forwardShape,
      annotations: destructive(),
    },
    toolHandler<ForwardArgs>(async ({ domain_id, host, url, frame }) => {
      const updated = await client.forwards.update(domain_id, host, {
        host,
        url,
        frame: frame ?? false,
      });
      return jsonResult(updated);
    }),
  );

  server.registerTool(
    "delete_forward",
    {
      description: "Delete an HTTP forward.",
      inputSchema: deleteForwardShape,
      annotations: destructive(),
    },
    toolHandler<DeleteForwardArgs>(async ({ domain_id, host }) => {
      await client.forwards.delete(domain_id, host);
      return textResult("Forward deleted.");
    }),
  );
}
